low-level/files/opt/seagate/sspl/setup/resource_map/server.py

```diff
@@ -22,28 +22,220 @@
+                "description": "",
+                "recommendation": "",
+                "specifics": []
+            }
+        }
+
+    def set_health_data(self, obj: dict, status, description=None,
+                        recommendation=None):
+        """Set the given or default health data as per health status."""
+        good_state = (status == "OK")
+        if not description:
+            description = "%s is %sin good state" % (
+                            obj.get("uid"),
+                            '' if good_state else 'not ')
+        if not recommendation:
+            recommendation = 'None' if good_state\
+                             else "Contact Seagate Support."
+
+        obj["health"].update({
+            "status": status,
+            "description": description,
+            "recommendation": recommendation
+        })
+
+    def get_sas_hba_info(self):
+        """Return the s SAS-HBA information."""
+        return self.get_sas_ports_info("sas_hba")
+
+    def get_sas_ports_info(self, leaf_node="sas_ports"):
+        """Return the s SAS ports information."""
+        sas_hba_data = []
+        health_data = []
+        for sas_rid in self.SAS_RESOURCE_ID:
+            r_data = self.get_data_template(sas_rid, False)
+            sas_ports_dict = self.sysfs.get_phy_negotiated_link_rate()
+            sas_ports_dict = sorted(sas_ports_dict.items(),
+                                    key=lambda kv: int(kv[0].split(':')[1]))
+            health = []
+            if sas_ports_dict:
+                for rid, rate in sas_ports_dict:
+                    r_data["health"]["specifics"].append({
+                        "resource_id": f"{sas_rid}:{rid.strip()}",
+                        "negotiated_link_rate": rate.strip()
+                    })
+                    if "Gbit".lower() in rate.strip().lower():
+                        health.append("OK")
+
+            # By some logic set health status, description, recommendation
+            health_data.append(health)
+            health_status = "OK" if "OK" in health else "Fault"
+            self.set_health_data(r_data, health_status)
+            if r_data["health"]["specifics"]:
+                sas_hba_data.append(r_data)
+
+        if leaf_node == "sas_hba":
+            return sas_hba_data
+
+        sas_ports_data = []
+        for heath_list in health_data:
+            for i, health in enumerate(heath_list):
+                if i % 4:
+                    continue
+
+                sas_port_data = self.get_data_template(f'sas_port_{i/4+1}',
+                                                       False)
+                sas_port_data["health"]["specifics"].append({
+                    "negotiated_link_rate": "12.0 Gbit"
+                })
+
+                health_status = "OK" if "OK" in heath_list[i:i+4] else "Fault"
+                self.set_health_data(sas_port_data, health_status)
+                sas_ports_data.append(sas_port_data)
```

Comment:
Add example of health data being generated in jira and give link here in review comment to visualize

Reply:
need hardware to get sas related data, since I don't have hw setup now, can't verify the sas related changes, since mock server does not simulate sas connections.

Reply:
Pasted the sample in jira

Reply:
yes, tried to add comments, function docstring and health view. pls mention if details needed somewhere.